### PR TITLE
Use interned Header Names for known headers not in the pre-allocated block

### LINKED ISSUE #31305: Large number of duplicate strings for concurrent websocket connections
Seems like something we could consider interning:

image

We could probably eliminate this by making these known headers (though I think we should potentially remove other known headers and replace them with the websocket headers 😄).

cc @benaadams

### PR DESCRIPTION
**Pull request by benaadams**
Use single instance strings for known header names; even if they aren't in the preserved header block.

This helps prevent multiple duplicates of the same string in long lived connections e.g. WebSockets

Resolves https://github.com/dotnet/aspnetcore/issues/31305

/cc @davidfowl

### CODE REVIEW

File: src/Http/Headers/src/HeaderNames.cs [hunk 1]
```diff
@@ -65,7 +65,7 @@ public static class HeaderNames
         public static readonly string Authorization = "Authorization";
 
         /// <summary>Gets the <c>baggage</c> HTTP header name.</summary>
-        public static readonly string Baggage = "baggage";
+        public static readonly string Baggage = "Baggage";
```
Comment (davidfowl) [inline src/Http/Headers/src/HeaderNames.cs:68]: Why this change?

File: src/Http/Headers/src/HeaderNames.cs [hunk 2]
```diff
@@ -245,10 +248,10 @@ public static class HeaderNames
         public static readonly string Translate = "Translate";
 
         /// <summary>Gets the <c>traceparent</c> HTTP header name.</summary>
-        public static readonly string TraceParent = "traceparent";
+        public static readonly string TraceParent = "TraceParent";
```
Comment (davidfowl) [inline src/Http/Headers/src/HeaderNames.cs:251]: Same question here?

File: src/Http/Headers/src/HeaderNames.cs [hunk 3]
```diff
@@ -245,10 +248,10 @@ public static class HeaderNames
...
 
         /// <summary>Gets the <c>tracestate</c> HTTP header name.</summary>
-        public static readonly string TraceState = "tracestate";
+        public static readonly string TraceState = "TraceState";
```
Comment (davidfowl) [inline src/Http/Headers/src/HeaderNames.cs:254]: Same here?

File: src/Http/Headers/src/HeaderNames.cs (hunk 1 above)
Comment (benaadams) [reply src/Http/Headers/src/HeaderNames.cs:68]: Is to match the casing in KnownHeaders; otherwise it lowercases all the variable names, is case insensitive so doesn't really matter, but the other way has more code changes https://github.com/dotnet/aspnetcore/blob/7dea0cb6736bc8ea2b53e5a716b926e7a80a4430/src/Servers/Kestrel/shared/KnownHeaders.cs#L112

File: src/Http/Headers/src/HeaderNames.cs (hunk 2 above)
Comment (benaadams) [reply src/Http/Headers/src/HeaderNames.cs:251]: As above
https://github.com/dotnet/aspnetcore/blob/7dea0cb6736bc8ea2b53e5a716b926e7a80a4430/src/Servers/Kestrel/shared/KnownHeaders.cs#L110

File: src/Http/Headers/src/HeaderNames.cs (hunk 3 above)
Comment (benaadams) [reply src/Http/Headers/src/HeaderNames.cs:254]: As above
https://github.com/dotnet/aspnetcore/blob/7dea0cb6736bc8ea2b53e5a716b926e7a80a4430/src/Servers/Kestrel/shared/KnownHeaders.cs#L111

**Comment by benaadams** [comment]
> Why this change?

Saw an; unrelated, issue in dotnet/runtime where someone was complaining about the casing of headers, so added change to keep baggage, tracestate, and traceparent 's previous casing

File: src/Middleware/WebSockets/src/HandshakeHelpers.cs [hunk 4]
```diff
@@ -43,11 +43,11 @@ public static bool CheckSupportedWebSocketRequest(string method, List<KeyValuePa
                 return false;
             }
 
-            foreach (var pair in headers)
+            foreach (var pair in interestingHeaders)
             {
                 if (string.Equals(HeaderNames.Connection, pair.Key, StringComparison.OrdinalIgnoreCase))
                 {
-                    if (string.Equals(Constants.Headers.ConnectionUpgrade, pair.Value, StringComparison.OrdinalIgnoreCase))
+                    if (string.Equals(HeaderNames.Upgrade, pair.Value, StringComparison.OrdinalIgnoreCase))
```
Comment (halter73) [inline src/Middleware/WebSockets/src/HandshakeHelpers.cs:50]: While this is a constant that should never change so technically correct, it feels wrong to repurpose the constant for `Upgrade: ` header name to compare against the value of a potential `Connection: Upgrade` header.

File: src/Middleware/WebSockets/src/HandshakeHelpers.cs [hunk 5]
```diff
@@ -72,12 +72,26 @@ public static bool CheckSupportedWebSocketRequest(string method, List<KeyValuePa
...
+                requestHeaders[HeaderNames.SecWebSocketVersion] = Constants.Headers.SupportedVersion;
+            }
+
             return validConnection && validUpgrade && validVersion && validKey;
         }
 
         public static void GenerateResponseHeaders(string key, string? subProtocol, IHeaderDictionary headers)
         {
-            headers[HeaderNames.Connection] = Constants.Headers.ConnectionUpgrade;
+            headers[HeaderNames.Connection] = HeaderNames.Upgrade;
```
Comment (halter73) [inline src/Middleware/WebSockets/src/HandshakeHelpers.cs:94]: ditto

File: src/Middleware/WebSockets/src/HandshakeHelpers.cs (hunk 4 above)
Comment (halter73) [reply src/Middleware/WebSockets/src/HandshakeHelpers.cs:50]: Actually thinking about the semantics of the WebSocket handshake more, I'm completely okay with it 😄

Comment (benaadams) [reply src/Middleware/WebSockets/src/HandshakeHelpers.cs:50]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Connection

`Upgrade` in the `Connection` header; refers to the `Upgrade` header; so it is technically correct to use it?

image

Comment (benaadams) [reply src/Middleware/WebSockets/src/HandshakeHelpers.cs:50]: Looks like `Connection` is a list of header names?

https://tools.ietf.org/html/rfc7230#section-6.1

> When a header field aside from Connection is used to supply control
   information for or about the current connection, the sender MUST list
   the corresponding field-name within the Connection header field.  A
   proxy or gateway MUST parse a received Connection header field before
   a message is forwarded and, for each connection-option in this field,
   remove any header field(s) from the message with the same name as the
   connection-option, and then remove the Connection header field itself
   (or replace it with the intermediary's own connection options for the
   forwarded message).
>
>   Hence, the Connection header field provides a declarative way of
   distinguishing header fields that are only intended for the immediate
   recipient ("hop-by-hop") from those fields that are intended for all
   recipients on the chain ("end-to-end"), enabling the message to be
   self-descriptive and allowing future connection-specific extensions
   to be deployed without fear that they will be blindly forwarded by
   older intermediaries.

File: src/Servers/Kestrel/Core/src/Internal/Http/HttpHeaders.cs [hunk 6]
```diff
@@ -126,6 +129,18 @@ public void Reset()
             ClearFast();
         }
 
+        protected static string GetInternedHeaderName(string name)
```
Comment (JamesNK) [inline src/Servers/Kestrel/Core/src/Internal/Http/HttpHeaders.cs:132]: This method looks like it will be used to set unknown headers.

Aren't most of the headers in this collection known? They will never be used because generated code will handle setting them, right?

Comment (benaadams) [reply src/Servers/Kestrel/Core/src/Internal/Http/HttpHeaders.cs:132]: There are 93 header names in `HeaderNames`; but only 20 in `commonHeaders` shared by `HttpRequestHeaders` and `HttpReponseHeaders`; could drop those 20; or have a different `HashSet<string>` for request and response but it doesn't seem worth it?

However yes, if the header is in the set for `HttpRequestHeaders` or `HttpReponseHeaders` and comes in on that direction it won't use the `HashSet<string>`, it only uses it for header names that aren't in that set rather than making a freshly constructed string long lived.

Any headers outside the set in `HeaderNames` will remain as the newly constructed string and won't be exchanged (i.e. the `HashSet<string>` is not updated while the app is running)

Comment (JamesNK) [reply src/Servers/Kestrel/Core/src/Internal/Http/HttpHeaders.cs:132]: Ok!

This two phases of hash lookups (intern hashset then unknown dictionary) makes me wonder whether it makes sense to write a custom collection.

As well as optimizing interning known keys, there is probably a opportunity to optimize memory usage. Most requests won't have that many unknown headers. e.g. an adaptive collection that can switch between `KeyValuePair<string, StringValues>[]` and `Dictionary<string, StringValues>` depending upon how many headers have been added.

Comment (benaadams) [reply src/Servers/Kestrel/Core/src/Internal/Http/HttpHeaders.cs:132]: > Any headers outside the set in `HeaderNames` will remain as the newly constructed string and won't be exchanged 

Is a question on whether to extend the values in `HeaderNames` and it becomes more valuable to do so with this change, rather than just being coinvent for the developer; e.g. missed standardise headers and defacto standard headers: `Prefer`, `Preference-Applied`, `X-Request-ID`, `X-Correlation-ID`, `X-Forwarded-For`, `X-Forwarded-Host`, `X-Forwarded-Proto`, `Refresh`, `X-Powered-By`, `Timing-Allow-Origin`

Though response set headers would likely be less important than request set headers; assuming they are being set by a single assembly so already an interned value.

However that's probably a separate issue...

File: src/Servers/Kestrel/shared/KnownHeaders.cs [hunk 7]
```diff
@@ -7,8 +7,11 @@
 using System.Globalization;
 using System.Linq;
 using System.Net.Http.HPack;
+using System.Reflection;
 using System.Text;
```
Comment (Tratcher) [inline src/Servers/Kestrel/shared/KnownHeaders.cs:12]:
```suggestion
```

File: src/Http/Headers/src/HeaderNames.cs [hunk 8]
```diff
@@ -274,10 +277,16 @@ public static class HeaderNames
         /// <summary>Gets the <c>WWW-Authenticate</c> HTTP header name.</summary>
         public static readonly string WWWAuthenticate = "WWW-Authenticate";
 
+        /// <summary>Gets the <c>X-Content-Type-Options</c> HTTP header name.</summary>
+        public static readonly string XContentTypeOptions = "X-Content-Type-Options";
```
[the remark this answers is not in the document]
Comment (benaadams) [reply src/Http/Headers/src/HeaderNames.cs:281]: > Why are you adding Link, X-Content-Type-Options and X-XSS-Protection here?

[www.webpagetest.org](https://www.webpagetest.org/) is one of the commonly used performance testing sites (along with Lighthouse based things for core web vitals e.g. [web.dev/measure](https://web.dev/measure/)); and it will call you out; rightly or wrongly, for not having `X-Content-Type-Options` and `X-XSS-Protection` set; amongst others:

Funnily `X-Content-Type-Options` is the only one [dotnet.microsoft.com](https://dotnet.microsoft.com/) does set:

image

image

vs

image

image

Just looks bad if you are explicitly called out for not setting them; but the Framework doesn't support them as a pre-defined headername when it has 90 other ones...

The `Link` header [rfc5988](https://tools.ietf.org/html/rfc5988); is a good one for using for preloading of fonts prior to the css being parsed and html starting to render to increase webpage performance and reduce [flash of unstyled content (FOUC)](https://en.wikipedia.org/wiki/Flash_of_unstyled_content) or flash of invisible text (FOIT) if you mark the text as unrenderable until the font is downloaded.

e.g. Adding this header on the landing HTML page

`link: <https://assets.ageofascent.net/fonts/teko-light-latin.woff2>; rel=preload; as=font; crossorigin=anonymous,<https://assets.ageofascent.net/fonts/pt_sans-web-regular.woff>; rel=preload; as=font; crossorigin=anonymous`

Chrome/Edge will then immediately start downloading the fonts; without yet having encountered the fonts in the parsing of the HTML (triggered by evaluating its css)

image

The browsers will also tell you off if you then don't use the preloaded font within the first few seconds of page load:

image

Anyway `Link` is one of the missed headers I use; so fringe benefit in adding for me with this change 😉

Its not adding it to the pre-reserved Kestrel headers; just to the `HeaderNames` constants; as otherwise they look fugly in use:

image

Though since you made me look; seems should add `"Referrer-Policy"` also 🤔

File: src/Servers/Kestrel/Core/src/Internal/Http/HttpHeaders.Generated.cs [hunk 9]
```diff
@@ -99,6 +99,106 @@ internal enum KnownHeaderType
         WWWAuthenticate,
     }
 
+    internal partial class HttpHeaders
+    {
+        protected readonly static HashSet<string> s_internedHeaderNames = new HashSet<string>(93, StringComparer.OrdinalIgnoreCase)
```
Comment (halter73) [inline src/Servers/Kestrel/Core/src/Internal/Http/HttpHeaders.Generated.cs:104]: Nit:

```suggestion
        private readonly static HashSet<string> s_internedHeaderNames = new HashSet<string>(93, StringComparer.OrdinalIgnoreCase)
```

Also, `s_` is a runtime convention, but I'm okay with it. It started with the shared code. It's only a matter of time 😆

**Review by halter73 — APPROVED** [review]

File: src/Middleware/WebSockets/src/HandshakeHelpers.cs [hunk 10]
```diff
@@ -34,7 +34,7 @@ internal static class HandshakeHelpers
         };
 
         // Verify Method, Upgrade, Connection, version,  key, etc..
-        public static bool CheckSupportedWebSocketRequest(string method, List<KeyValuePair<string, string>> headers)
+        public static bool CheckSupportedWebSocketRequest(string method, List<KeyValuePair<string, string>> interestingHeaders, IHeaderDictionary requestHeaders)
```
[the remark this answers is not in the document]
Comment (benaadams) [reply src/Middleware/WebSockets/src/HandshakeHelpers.cs:37]: Will do this in a follow up?

[the remark this answers is not in the document]
Comment (benaadams) [reply src/Middleware/WebSockets/src/HandshakeHelpers.cs:37]: https://github.com/dotnet/aspnetcore/issues/31373

File: src/Servers/Kestrel/Core/src/Internal/Http/HttpHeaders.cs (hunk 6 above)
Comment (benaadams) [reply src/Servers/Kestrel/Core/src/Internal/Http/HttpHeaders.cs:132]: Added issue for other headers https://github.com/dotnet/aspnetcore/issues/31374
